### PR TITLE
Add MoE router similarity and expert fraction distillation metrics

### PR DESCRIPTION
**Pull request by JamesDeng42**
# Description

This PR introduces new metrics for evaluating the quality of Mixture-of-Experts (MoE) distillation. When distilling an MoE teacher into an MoE student, it is critical to understand not just the KL divergence of the final logits, but whether the student's routing decisions align with the teacher's, and how evenly the tokens are distributed across experts.

Key Features:
   * Expert Fraction Distribution: Automatically computes and logs distill/student_expert_{i}_fraction and distill/teacher_expert_{i}_fraction whenever a model has num_experts > 1.
   * Router Similarity: Introduces a new configuration flag record_router_similarity_metrics. When enabled, it computes per-layer Top-1 and Top-K routing overlap between the student and teacher models (e.g.,
     distill/layer_0_router_similarity_top1).
   * Eval Logging: Updated the distillation evaluation loop to conditionally run a stop_gradient forward pass of the teacher model so these metrics are populated during eval steps.

# Tests

Added test_router_similarity_metrics to verify that Top-1 and Top-K similarity ratios correctly ignore padded tokens (pad_id=0) and broadcast properly.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -2298,6 +2298,10 @@ def __call__(
     routing_inputs = inputs if gate_inputs is None else gate_inputs.astype(gate_dtype)
     gate_logits, pre_bias_logits = self.gate(routing_inputs)
 
+    if getattr(self.config, "record_router_similarity_metrics", False):
```
Comment (vlad-karp) [inline src/maxtext/layers/moe.py:2301]: can we use a more readable self.config.record_router_similarity_metrics instead?

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 2]
```diff
@@ -377,6 +383,8 @@ def __init__(
     self.alpha = alpha
     self.beta_feature = beta_feature
     self.layer_indices = jnp.array(layer_indices) if layer_indices is not None else None
+    self.num_experts = num_experts
```
Comment (vlad-karp) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py:386]: why do we need num_experts besides just record_router_similarity_metrics?

**Review by vlad-karp — APPROVED** [review]
